frb_codegen/src/commands.rs

```diff
@@ -18,11 +19,53 @@ fn call_shell(cmd: &str) -> Output {
     execute_command("sh", &["-c", cmd], None)
 }
 
-pub fn ensure_tools_available() -> Result {
-    let output = call_shell("dart pub global list");
-    let output = String::from_utf8_lossy(&output.stdout);
-    if !output.contains("ffigen") {
-        return Err(Error::MissingExe(String::from("ffigen")));
+pub fn ensure_tools_available(maybe_dart_root: &Option<String>) -> Result {
+    let dart_root = maybe_dart_root
+        .as_deref()
+        .unwrap_or(env!("CARGO_MANIFEST_DIR").into());
+    let toolchain = guess_toolchain(&dart_root).unwrap();
+    if toolchain == DartToolchain::Dart {
+        if !call_shell("dart --version").status.success() {
+            return Err(Error::MissingExe(String::from("dart")));
+        };
+    } else {
+        if !call_shell("flutter --version").status.success() {
+            return Err(Error::MissingExe(String::from("flutter")));
+        };
+    };
+
+    let ffi = ensure_dependencies(&dart_root, "ffi").unwrap();
+    if let Some(ffi) = ffi {
+        if VersionReq::parse("^2.0.1").unwrap() != ffi {
+            return Err(Error::InvalidDep(
+                "ffi".into(),
+                "dependencies".into(),
+                "^2.0.1".into(),
+            ));
+        }
+    } else {
+        return Err(Error::MissingDep(
+            "ffi".into(),
+            "dependencies".into(),
+            "^2.0.1".into(),
+        ));
+    }
+
+    let ffigen = ensure_dev_dependencies(&dart_root, "ffigen").unwrap();
+    if let Some(ffigen) = ffigen {
+        if VersionReq::parse("^6.0.1").unwrap() != ffigen {
+            return Err(Error::InvalidDep(
+                "ffigen".into(),
+                "dev_dependencies".into(),
+                "^6.0.1".into(),
+            ));
+        }
+    } else {
+        return Err(Error::MissingDep(
+            "ffigen".into(),
+            "dev_dependencies".into(),
+            "^6.0.1".into(),
+        ));
```

Comment:
nit: Seems to have duplicated code here. Maybe we can extract such "ensure dependency or return error" pattern into a separate function

Reply:
yes and as a reminder to myself I also have to check my comparisons e.g. `VersionReq::parse("^6.0.1").unwrap() != ffigen` because I'm pretty sure this needs to be improved.

Reply:
so here I decided to drop package version comparison for now (might reintroduce in a later PR) because it turns out there's some issue when trying to compare `VersionReq` with `VersionReq`, as discussed [there](https://github.com/dtolnay/semver/pull/223).
a quick test can allow you to assert it :
```rust
#[test]
fn assert_semver_here() {
    // ok
    let given = VersionReq::from_str(">=2.0.1, <3.0.0").unwrap();
    let found = VersionReq::from_str(">=2.0.1, <3.0.0").unwrap();
    assert_eq!(given,found);
    
    // ok
    let given = VersionReq::from_str("~2.0.1").unwrap();
    let found = VersionReq::from_str("~2.0.1").unwrap();
    assert_eq!(given,found);
    
    // weird
    let given = VersionReq::from_str(">=2.0.1, <2.1.0").unwrap();
    let found = VersionReq::from_str("~2.0.1").unwrap();
    assert_ne!(given,found);
}
```
According to [Cargo tilde requirements](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#tilde-requirements), the last test above should pass.

Also please note there's a difference in semantics between [Dart caret syntax](https://dart.dev/tools/pub/dependencies#caret-syntax) and [Rust caret requirements](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#caret-requirements).

Reply:
Btw I have never seen `~` in Dart, only see `^` and `>=,<` yet. So no worries about that case.

Reply:
Yes, the complication arises because [VersionReq](https://docs.rs/semver/latest/semver/struct.VersionReq.html) in `semver` crate abides by Cargo semantics.

But Cargo and Dart semantics differ, for example:
* as of Cargo's, `^1.2.3` literally means `1.2.3` (see [Cargo caret requirements](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#caret-requirements))
* whereas, as of Dart's, `^1.2.3` instead means `>=1.2.3 <2.0.0` (see [Dart caret syntax](https://dart.dev/tools/pub/dependencies#caret-syntax))

It gets even weirder when Dart caret syntax also states that `^0.1.2` means `>=0.1.2 <0.2.0`.
You can see from the former example that caret sometimes goes up to the next **major** (exclusive), and from this example that it can go up to the next **minor** (exclusive).

In Cargo, the closest semantic to define such range is tilde, but it's not 100% interchangeable, for example:
* `~1.2.3` means `>=1.2.3, <1.3.0` (remember how in Dart `^1.2.3` meant `>=1.2.3 <2.0.0`)

As already mentioned earlier, it gets even more complicated when trying to compare 2 `VersionReq`s together (which I'd need to provide user with a reliable information), as stated in [this unmerged PR](https://github.com/dtolnay/semver/pull/223).

That's why all in all, instead of providing a half-baked version detection, I think it's more reasonable to let it for a feature PR.